### DISCUSSION TITLE
Converting what you get from auxvariables/auxkernels into material objects/BCs calculation?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#18774] [Q&A General] [answered]
Hi guys

How do you convert what you get from auxvariables/auxkernels into material objects/ BCs?

I want to use the minimum length obtained from `ElementLengthAux` to calculate drainage (PorousFlowSink), is there any example, where should I start?

I want to add it here (`PorousFlowSink.C`)
```
Real
PorousFlowSinkTC::computeQpResidual()
{
  Real flux = _test[_i][_qp] * multiplier();
  if (_use_mobility)
  {
    const Real k =
        ((*_permeability)[_qp] * _normals[_qp]) * _normals[_qp]; // do not upwind permeability
    flux *= (*_fluid_density_node)[_i][_ph] * k / (*_fluid_viscosity)[_i][_ph]; ***********HERE************
  }
```

Kind regards,
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Hello

I dont know about that file in particular but you can use either or both material properties and auxiliary variables in boundary conditions. 
This object already seems to use material properties so you can look at it for its use of the MaterialPropertyInterface. For coupling in auxiliary variables you can look at the CoupledForce kernel and its use of the Coupleable interface.

Guillaume

**Reply by Traiwit** [reply]
Hi @GiudGiud, which would be easier?

1. Auxvariable > material > BCs
2. Auxvariable > BCs

I'm still not sure how to call what I got from `ElementLengthAux` to `PorousFlowSink.C`

The example below is how density and viscosity (material object) are called

```
    _fluid_density_node(_has_mobility ? &getMaterialProperty<std::vector<Real>>(
                                            "PorousFlow_fluid_phase_density_nodal")
                                      : nullptr),
    _dfluid_density_node_dvar(_has_mobility ? &getMaterialProperty<std::vector<std::vector<Real>>>(
                                                  "dPorousFlow_fluid_phase_density_nodal_dvar")
                                            : nullptr),
    _fluid_viscosity(_has_mobility
                         ? &getMaterialProperty<std::vector<Real>>("PorousFlow_viscosity_nodal")
                         : nullptr),
    _dfluid_viscosity_dvar(_has_mobility ? &getMaterialProperty<std::vector<std::vector<Real>>>(
                                               "dPorousFlow_viscosity_nodal_dvar")
                                         : nullptr),
```

is there something like getAuxvariable? also is there a documentary on this stuff? I think I should read more about MOOSE .C and .h file before I start doing stuff like this.

Thank you

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
Auxvariable > BCs for sure.
For AuxVariable there is the Coupleable interface. Here's the constructor from CoupledForce (a kernel) where an auxvariable can be coupled in:
```
CoupledForce::CoupledForce(const InputParameters & parameters)
  : Kernel(parameters), _v_var(coupled("v")), **_v(coupledValue("v")),** _coef(getParam<Real>("coef"))
{
  if (_var.number() == _v_var)
    mooseError("Coupled variable 'v' needs to be different from 'variable' with CoupledForce, "
               "consider using the CoefReaction kernel or something similar");
}
```

then the value of the coupled variable (here _v) is obtained with `_v[_qp]`

The variable can also be obtained (instead of the value) using the coupled() routine, but i dont think you need this here.

**Reply by GiudGiud** [reply]
Are you retrieving the aux variable from the same domain the boundary condition is applied in? 
Or does it 'live' on the other side?

**Reply by Traiwit** [reply]
@GiudGiud Thank you very much. The auxvariable is obtained from the same domain.

The original PorousFlowSink equation is basically darcy's law, but I want to use Abaqus equation for the drainage instead
![image](https://user-images.githubusercontent.com/79562439/132167320-3bd21db9-086f-4f64-b6ec-5f6a46f13942.png)
https://abaqus-docs.mit.edu/2017/English/SIMACAEPRCRefMap/simaprc-c-porousflow.htm

Apparently, the c can be calculated using modified `ElementLengthAux`

Anyways, I will look into this further by calling auxvariable via  `_v(coupledValue("v")` then add `_v[_qp]` into the equation I mentioned above. I will keep you updated, I really appreciate all your help

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
it's working now, thank you! @GiudGiud
